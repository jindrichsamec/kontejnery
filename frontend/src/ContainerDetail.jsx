import React, { Component, PropTypes } from 'react'
import { Modal, Button } from 'react-bootstrap'
import { observable } from 'mobx'
import { observer } from 'mobx-react'
import Icon from './Icon'
import { formatDate, formatDayName, formatTimeInterval } from './utils/DateInterval'
import SearchQuery from './model/SearchQuery'

export default observer(class ContainerDetail extends Component {

  static propTypes = {
    id: PropTypes.number.isRequired,
    name: PropTypes.string.isRequired,
    onClose: PropTypes.func.isRequired
  }

  model = observable({
    detailInfo: null
  })

  componentDidMount() {
    const since = SearchQuery.since.toISOString();
    fetch(`${process.env.REACT_APP_API_HOST}/api/${this.props.id}?since=${since}`).then(this.handleSuccess, this.handleFail)
  }

  normalizeData = (data) => {
    const terms = data.terms.map((item) => {
      return {
        ...item,
        since: new Date(item.since),
        till: new Date(item.till)
      }
    })
    return {
      ...data,
      terms
    }
  }

  handleSuccess = (response) => {
    response.json().then(json => {
      const data = this.normalizeData(json.data)
      this.setDetailInfo(data)
    })
  }

  handleFail = () => {
    console.error('Error during the request');
  }

  setDetailInfo = (info) => {
    this.model.detailInfo = info
  }

  renderTerms(info) {
    return <table className="table table-stripped">
        <thead>
          <tr>
            <th>Den</th>
            <th>Datum</th>
            <th>Čas přistavení</th>
          </tr>
        </thead>
        <tbody>
          {info.terms.map((term) => {
            return <tr key={term.id}>
              <td>{formatDayName(term.since.getDay())}</td>
              <td>{formatDate(term.since)}</td>
              <td>{formatTimeInterval(term.since, term.till)}</td>
            </tr>
          })}
        </tbody>
      </table>
  }

  renderInfo() {
    return <div>
      <div className="bs-callout bs-callout-info small help-block">
        <strong>Do kontejneru patří</strong>
        <p>starý nábytek, koberce a linolea, zrcadla, umyvadla, vany a WC mísy, staré sportovní náčiní, autosklo a kovové předměty.</p>
      </div>
      <div className="bs-callout bs-callout-danger small help-block">
        <strong>Do kontejneru nepatří</strong>
        <p>odkládat živnostenský odpad, nebezpečný odpad (např.: autobaterie, zářivky, barvy, rozpouštědla, motorové oleje a obaly od nich), bioodpad, stavební odpad, dále pak pneumatiky, elektrospotřebiče, televizory a PC monitory, počítače, lednice, mrazáky a sporáky.</p>
      </div>
      <p className="small">
        Zdroj: <a href="https://www.praha8.cz/kontejnery-na-velkoobjemovy-odpad.html" title="MČ Praha 8: Velkoobjemové kontejnery - Kontejnery na velkoobjemový odpad">
          https://www.praha8.cz/kontejnery-na-velkoobjemovy-odpad.html
        </a>
      </p>
    </div>;

  }

  render() {
    const { onClose, name } = this.props;
    const { detailInfo } = this.model

    return (<Modal show={true} onHide={onClose}>
      <Modal.Header closeButton>
        <Modal.Title>{name}</Modal.Title>
      </Modal.Header>
      <Modal.Body>
        {detailInfo === null ? <Icon name="circle-o-notch" /> : this.renderTerms(detailInfo)}
        {detailInfo && this.renderInfo()}
      </Modal.Body>
      <Modal.Footer>
        <Button className="btn btn-default" onClick={onClose}>Zavřít</Button>
      </Modal.Footer>
    </Modal>);
  }

})
