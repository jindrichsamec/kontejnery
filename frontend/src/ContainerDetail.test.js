import ContainerDetail from './ContainerDetail'

it('renders without crashing', () => {
  const props = {
    location: {state: {name:"Test"}},
    match: {params: {id: 4}}
  }
  const wrapper = shallow(<ContainerDetail {...props} />);
  expect(wrapper).toMatchSnapshot();
});
