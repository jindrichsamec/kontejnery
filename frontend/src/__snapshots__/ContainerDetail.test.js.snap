exports[`test renders without crashing 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <Modal
    animation={true}
    autoFocus={true}
    backdrop={true}
    bsClass="modal"
    dialogComponentClass={[Function]}
    enforceFocus={true}
    keyboard={true}
    manager={
        ModalManager {
            "containers": Array [],
            "data": Array [],
            "handleContainerOverflow": true,
            "hideSiblingNodes": true,
            "modals": Array [],
          }
    }
    onHide={[Function]}
    renderBackdrop={[Function]}
    restoreFocus={true}
    show={true}>
    <ModalHeader
        aria-label="Close"
        bsClass="modal-header"
        closeButton={true}>
        <ModalTitle
            bsClass="modal-title"
            componentClass="h4">
            Test
        </ModalTitle>
    </ModalHeader>
    <ModalBody
        bsClass="modal-body"
        componentClass="div">
        <Icon
            name="circle-o-notch" />
    </ModalBody>
    <ModalFooter
        bsClass="modal-footer"
        componentClass="div">
        <Button
            active={false}
            block={false}
            bsClass="btn"
            bsStyle="default"
            className="btn btn-default"
            disabled={false}
            onClick={[Function]}>
            Zavřít
        </Button>
    </ModalFooter>
</Modal>,
  "nodes": Array [
    <Modal
      animation={true}
      autoFocus={true}
      backdrop={true}
      bsClass="modal"
      dialogComponentClass={[Function]}
      enforceFocus={true}
      keyboard={true}
      manager={
            ModalManager {
                  "containers": Array [],
                  "data": Array [],
                  "handleContainerOverflow": true,
                  "hideSiblingNodes": true,
                  "modals": Array [],
                }
      }
      onHide={[Function]}
      renderBackdrop={[Function]}
      restoreFocus={true}
      show={true}>
      <ModalHeader
            aria-label="Close"
            bsClass="modal-header"
            closeButton={true}>
            <ModalTitle
                  bsClass="modal-title"
                  componentClass="h4">
                  Test
            </ModalTitle>
      </ModalHeader>
      <ModalBody
            bsClass="modal-body"
            componentClass="div">
            <Icon
                  name="circle-o-notch" />
      </ModalBody>
      <ModalFooter
            bsClass="modal-footer"
            componentClass="div">
            <Button
                  active={false}
                  block={false}
                  bsClass="btn"
                  bsStyle="default"
                  className="btn btn-default"
                  disabled={false}
                  onClick={[Function]}>
                  Zavřít
            </Button>
      </ModalFooter>
</Modal>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <ContainerDetail
        location={
                Object {
                        "state": Object {
                          "name": "Test",
                        },
                      }
        }
        match={
                Object {
                        "params": Object {
                          "id": 4,
                        },
                      }
        } />,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": ContainerDetail {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "handleClose": [Function],
        "handleFail": [Function],
        "handleSuccess": [Function],
        "model": Object {
          "detailInfo": null,
        },
        "normalizeData": [Function],
        "props": Object {
          "location": Object {
            "state": Object {
              "name": "Test",
            },
          },
          "match": Object {
            "params": Object {
              "id": 4,
            },
          },
        },
        "refs": Object {},
        "render": [Function],
        "setDetailInfo": [Function],
        "state": null,
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <Modal
          animation={true}
          autoFocus={true}
          backdrop={true}
          bsClass="modal"
          dialogComponentClass={[Function]}
          enforceFocus={true}
          keyboard={true}
          manager={
                    ModalManager {
                              "containers": Array [],
                              "data": Array [],
                              "handleContainerOverflow": true,
                              "hideSiblingNodes": true,
                              "modals": Array [],
                            }
          }
          onHide={[Function]}
          renderBackdrop={[Function]}
          restoreFocus={true}
          show={true}>
          <ModalHeader
                    aria-label="Close"
                    bsClass="modal-header"
                    closeButton={true}>
                    <ModalTitle
                              bsClass="modal-title"
                              componentClass="h4">
                              Test
                    </ModalTitle>
          </ModalHeader>
          <ModalBody
                    bsClass="modal-body"
                    componentClass="div">
                    <Icon
                              name="circle-o-notch" />
          </ModalBody>
          <ModalFooter
                    bsClass="modal-footer"
                    componentClass="div">
                    <Button
                              active={false}
                              block={false}
                              bsClass="btn"
                              bsStyle="default"
                              className="btn btn-default"
                              disabled={false}
                              onClick={[Function]}>
                              Zavřít
                    </Button>
          </ModalFooter>
</Modal>,
        "_debugID": 2,
        "_renderedOutput": <Modal
          animation={true}
          autoFocus={true}
          backdrop={true}
          bsClass="modal"
          dialogComponentClass={[Function]}
          enforceFocus={true}
          keyboard={true}
          manager={
                    ModalManager {
                              "containers": Array [],
                              "data": Array [],
                              "handleContainerOverflow": true,
                              "hideSiblingNodes": true,
                              "modals": Array [],
                            }
          }
          onHide={[Function]}
          renderBackdrop={[Function]}
          restoreFocus={true}
          show={true}>
          <ModalHeader
                    aria-label="Close"
                    bsClass="modal-header"
                    closeButton={true}>
                    <ModalTitle
                              bsClass="modal-title"
                              componentClass="h4">
                              Test
                    </ModalTitle>
          </ModalHeader>
          <ModalBody
                    bsClass="modal-body"
                    componentClass="div">
                    <Icon
                              name="circle-o-notch" />
          </ModalBody>
          <ModalFooter
                    bsClass="modal-footer"
                    componentClass="div">
                    <Button
                              active={false}
                              block={false}
                              bsClass="btn"
                              bsStyle="default"
                              className="btn btn-default"
                              disabled={false}
                              onClick={[Function]}>
                              Zavřít
                    </Button>
          </ModalFooter>
</Modal>,
      },
      "_renderedNodeType": 1,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <ContainerDetail
    location={
        Object {
            "state": Object {
              "name": "Test",
            },
          }
    }
    match={
        Object {
            "params": Object {
              "id": 4,
            },
          }
    } />,
}
`;
